katdal/chunkstore_s3.py

```diff
@@ -398,35 +438,159 @@ class S3ChunkStore(ChunkStore):
+        chunk_name : str, optional
+            Name of chunk, used for error reporting only
+        ignored_errors : collection of int, optional
+            HTTP status codes that are treated like 200 OK, not raising an error
+        timeout : tuple or float, optional
+            Override timeout for this request (use the store timeout by default)
+        kwargs : optional
+            These are passed on to :meth:`requests.Session.request`
+
+        Yields
+        ------
+        response : `requests.Response`
+            HTTP response, considered successful
+
+        Raises
+        ------
+        AuthorisationFailed
+            If the request is not authorised by appropriate token or credentials
+        S3ObjectNotFound
+            If S3 object request fails because it does not exist
+        S3ServerGlitch
+            If HTTP error has a retryable status code (i.e. temporary glitch)
+        StoreUnavailable
+            If a general HTTP error occurred that is not ignored
+        """
+        kwargs['timeout'] = self.timeout if timeout is () else timeout
+        # Use _standard_errors to filter errors emanating from within with-block
         with self._standard_errors(chunk_name), self._session_pool() as session:
-            with session.request(method, url, *args, **kwargs) as response:
-                _raise_for_status(response)
-                yield response
+            with session.request(method, url, **kwargs) as response:
+                # Turn error responses into the appropriate exception, like raise_for_status
+                status = response.status_code
+                prefix = 'Chunk {!r}: '.format(chunk_name) if chunk_name else ''
+                msg = '{}Store responded with HTTP error {} ({}) to request: {} {}'.format(
+                    prefix, status, response.reason, response.request.method, response.url)
+                if status == 401:
+                    raise AuthorisationFailed(msg)
+                elif status in (403, 404):
+                    # Ceph RGW returns 403 for missing keys due to our bucket policy
+                    # (see https://tracker.ceph.com/issues/38638 for discussion)
+                    raise S3ObjectNotFound(msg)
+                elif self._retries.is_retry(method, status):
+                    raise S3ServerGlitch(msg, status)
+                elif 400 <= status < 600 and status not in ignored_errors:
+                    raise StoreUnavailable(msg)
+                else:
+                    try:
+                        yield response
+                    except TruncatedRead as trunc_error:
+                        # A truncated read is considered a glitch with custom status
+                        glitch_error = S3ServerGlitch(prefix + str(trunc_error),
+                                                      _TRUNCATED_HTTP_STATUS_CODE)
+                        raise_from(glitch_error, trunc_error)
```

Comment:
I'm trying to reuse the urllib3 Retry internals as much as possible, including its backoff strategy. I get that by pretending that truncated reads are status responses, and then I need to make up a status code when calling `retries.increment` to process the retry. I figured I'll keep that unique instead of e.g. picking the the first status code in the glitch list.

I had several iterations of e.g. catching `TruncatedRead` along with `S3ServerGlitch` in `complete_request` before settling on the current code. This seemed like the cleanest option in the end (even though it's a bit abusive of HTTP codes).

Reply:
I guess that's okay, provided that the made-up number isn't shown to the user, leading to a confused search for what this HTTP status code is.

Reply:
That won't happen because the status error reports happen _before_ the content is read, and the truncation happens during read (with a different error message).